---
title: "R Notebook"
output: html_notebook
---


# Raw 16S rRNA Gene Sequencing Data to ASVs

```{r}

library(DECIPHER); packageVersion("DECIPHER")
library(dada2); packageVersion("dada2")
library(tidyverse)
library(ggtext)

```


```{r}
fnFs <- sort(list.files("raw_fastq", pattern="_R1.fastq", full.names = TRUE))
fnRs <- sort(list.files("raw_fastq", pattern="_R2.fastq", full.names = TRUE))
# Extract sample names, assuming filenames have format: SAMPLENAME_XXX.fastq
sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 1)

```

```{r}
# Place filtered files in filtered/ subdirectory
filtFs <- file.path("filtered_fastq", paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path("filtered_fastq", paste0(sample.names, "_R_filt.fastq.gz"))
names(filtFs) <- sample.names
names(filtRs) <- sample.names

```

```{r}
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(240,160), maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE, compress=TRUE, multithread=TRUE)

```

```{r}
errF <- learnErrors(filtFs, multithread=TRUE)
errR <- learnErrors(filtRs, multithread=TRUE)

```

```{r}
dadaFs <- dada(filtFs, err=errF, multithread=TRUE, pool=TRUE)
dadaRs <- dada(filtRs, err=errR, multithread=TRUE, pool=TRUE)

```


```{r}
mergers <- mergePairs(dadaFs, filtFs, dadaRs, filtRs, verbose=TRUE)

```

```{r}
seqtab <- makeSequenceTable(mergers)
dim(seqtab)

```

```{r}
# Inspect distribution of sequence lengths
table(nchar(getSequences(seqtab)))

```

```{r}
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)

```

```{r}
print(dim(seqtab.nochim))
print(sum(seqtab.nochim))
sum(seqtab.nochim)/sum(seqtab)

# 2273 ASVs after chimera removal -- includes all samples (Negative and positive controls)
# The frequency of chimeric sequences varies substantially from dataset to dataset, and depends on on factors including experimental procedures and sample complexity. Here chimeras make up about 60% of the merged sequence variants, but when we account for the abundances of those variants we see they account for only about 7% of the merged sequence reads.

```

```{r}
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))
# If processing a single sample, remove the sapply calls: e.g. replace sapply(dadaFs, getN) with getN(dadaFs)
colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim")
rownames(track) <- sample.names
track %>% 
  as_tibble(rownames = "SampleID")

```



```{r}
dna <- DNAStringSet(getSequences(seqtab.nochim))
load("GTDB_r202-mod_April2021.RData")
ids <- IdTaxa(dna, trainingSet, strand="top", processors=NULL, verbose=FALSE)


```

```{r}
rownames(seqtab.nochim) <- sapply(strsplit(rownames(seqtab.nochim), "_"), `[`, 1)

# write all results first then filter to remove ASVs and samples
# giving our seq headers more manageable names (ASV_1, ASV_2...)
asv_seqs <- colnames(seqtab.nochim)
asv_headers <- vector(dim(seqtab.nochim)[2], mode="character")

for (i in 1:dim(seqtab.nochim)[2]) {
  asv_headers[i] <- paste(">ASV", i, sep="_")
}

# making and writing out a fasta of our final ASV seqs:
asv_fasta <- c(rbind(asv_headers, asv_seqs))
write(asv_fasta, "ASVs_all.fa")



# count table:
asv_tab <- t(seqtab.nochim)
row.names(asv_tab) <- sub(">", "", asv_headers)

# tax table:
# creating table of taxonomy and setting any that are unclassified as "NA"
ranks <- c("domain", "phylum", "class", "order", "family", "genus", "species")
asv_tax <- t(sapply(ids, function(x) {
  m <- match(ranks, x$rank)
  taxa <- x$taxon[m]
  taxa[startsWith(taxa, "unclassified_")] <- NA
  taxa
}))
colnames(asv_tax) <- ranks
rownames(asv_tax) <- gsub(pattern=">", replacement="", x=asv_headers)


# remove ASVs not Bacteria or Mito - dont need these for any samples, even controls
asv_tax_keep <- asv_tax %>% 
  as_tibble(rownames = "ASV") %>% 
  filter(domain == "Bacteria") %>% 
  filter(!genus %in% c("Mitochondria_genus_incertae_sedis"))

asv_tax_keep %>% 
    write_tsv("ASVs_all_taxonomy.tsv")

asv_tab_keep <- asv_tab %>% 
  as_tibble(rownames = "ASV") %>% 
  filter(ASV %in% asv_tax_keep$ASV)

asv_tab_keep %>% 
  write_tsv("ASVs_all_counts.tsv")



```

```{r}
library(tidyverse)

asv_tax_keep <- read_tsv("ASVs_all_taxonomy.tsv")
asv_tab_keep <- read_tsv("ASVs_all_counts.tsv")
                    
# remove samples dont need and ASVs where sum == 0 as a result
# remove sample with low total read count too - S50

asv_tab_keep_samples <- asv_tab_keep %>% 
  select(-`Extraction-ctrl`, -`Mock-community`, -`No-DNA`, -`S50`) %>% 
  filter_if(is.numeric, any_vars(. != 0 & !is.na(.)))

asv_tab_keep_samples %>%
  write_tsv("ASVs_filtered_counts.tsv")


# count sequences per sample
asv_tab_keep_samples %>% 
  pivot_longer(cols = !ASV, names_to = "SampleID", values_to = "Counts") %>% 
  group_by(SampleID) %>% 
  summarise(Reads = sum(Counts)) %>% 
  arrange(desc(Reads)) %>% 
  write_tsv("sample_counts_after_asv_filter.tsv")


```




# Differentially abundant ASVs

```{r}
library(ANCOMBC)
library(phyloseq)
library(microbiome)
library(tidyverse)
library(DT)
library(metagMisc)
library(ggtext)

```


```{r}

# Since not interested in running all pariwise comparisons, will correct for multiple comparisons after running each comparison of interest. Merge output tables then correct for multiple comparisons


otumat <- read.table("ASVs_filtered_counts.tsv" , sep = "\t", header = TRUE, row.names = 1) %>% 
  select(sort(current_vars())) %>% 
  as.matrix()

taxmat <- read.table("ASVs_taxonomy_filtered.tsv", sep = "\t", header = TRUE, row.names = 1) %>% 
  select(sort(current_vars())) %>% 
  as.matrix()

df_taxa <- read_tsv("ASVs_taxonomy_filtered.tsv") 

df_sample <- read.table("metadata_noabx.tsv", header = TRUE, row.names = 1)
df_sample <- df_sample[ order(row.names(df_sample)), ]
df_sample <- df_sample %>% 
  rename(sample_names = sample) %>% 
  unite("site_necropsy", site, necropsy, sep = "-", remove = FALSE)

OTU = otu_table(otumat, taxa_are_rows = TRUE)
TAX = tax_table(taxmat)
sampledata = sample_data(df_sample)
physeq = phyloseq(OTU, TAX, sampledata)




```

```{r}
# -Jej_5wo vs Jej_7wo

physeq_subset <- physeq %>% 
  subset_samples(diet == "NoAbx" & site == "Jej") 

physeq_subset <- prune_taxa(taxa_sums(physeq_subset) > 0, physeq_subset) 

#physeq_subset <- phyloseq_filter_prevalence(physeq_subset, prev.trh = 0.0, abund.trh = 5, threshold_condition = "AND", abund.type = "mean")

physeq_subset

phyloseq_to_df(physeq_subset, addtax = F) %>% 
  pivot_longer(!OTU, names_to = "sample_names", values_to = "counts") %>% 
  left_join(df_sample, by = "sample_names") %>% 
  select(sample_names, site_necropsy) %>% 
  distinct(site_necropsy)


out <- ancombc(phyloseq = physeq_subset, formula = "site_necropsy", 
              p_adj_method = "fdr", zero_cut = 0.90, lib_cut = 0, 
              struc_zero = FALSE, neg_lb = TRUE, tol = 1e-5, 
              max_iter = 100, conserve = FALSE, alpha = 0.05, global = FALSE)

res <- out$res

res_beta <- res$beta %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "beta")

res_se <- res$se %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "se")

res_w <- res$W %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "W")

res_pval <- res$p_val %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "p_val")

res_qval <- res$q_val %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "q_val")

res_diff <- res$diff_abn %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "diff_abn")

res_long <- res_beta %>% 
  left_join(res_se, by = c("ASV", "comparison")) %>% 
  left_join(res_w, by = c("ASV", "comparison")) %>% 
  left_join(res_pval, by = c("ASV", "comparison")) %>% 
  #left_join(res_qval, by = c("ASV", "comparison")) %>% 
  #left_join(res_diff, by = c("ASV", "comparison")) %>% 
  left_join(df_taxa, by = "ASV") %>% 
  mutate(comparison = "Jej_5wo-vs-Jej_7wo") %>% 
  mutate(reference_group = "Jej_7wo") %>% 
  mutate(comparison_group = "Jej_5wo")  
  
```




```{r}
# Il_5wo vs Il_7wo

physeq_subset <- physeq %>% 
  subset_samples(diet == "NoAbx" & site == "Il") 

physeq_subset <- prune_taxa(taxa_sums(physeq_subset) > 0, physeq_subset) 

#physeq_subset <- phyloseq_filter_prevalence(physeq_subset, prev.trh = 0.0, abund.trh = 5, threshold_condition = "AND", abund.type = "mean")

physeq_subset

phyloseq_to_df(physeq_subset, addtax = F) %>% 
  pivot_longer(!OTU, names_to = "sample_names", values_to = "counts") %>% 
  left_join(df_sample, by = "sample_names") %>% 
  select(sample_names, site_necropsy) %>% 
  distinct(site_necropsy)


out <- ancombc(phyloseq = physeq_subset, formula = "site_necropsy", 
              p_adj_method = "fdr", zero_cut = 0.90, lib_cut = 0, 
              struc_zero = FALSE, neg_lb = TRUE, tol = 1e-5, 
              max_iter = 100, conserve = FALSE, alpha = 0.05, global = FALSE)

res <- out$res

res_beta <- res$beta %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "beta")

res_se <- res$se %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "se")

res_w <- res$W %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "W")

res_pval <- res$p_val %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "p_val")

res_qval <- res$q_val %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "q_val")

res_diff <- res$diff_abn %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "diff_abn")

res_long <- res_beta %>% 
  left_join(res_se, by = c("ASV", "comparison")) %>% 
  left_join(res_w, by = c("ASV", "comparison")) %>% 
  left_join(res_pval, by = c("ASV", "comparison")) %>% 
  #left_join(res_qval, by = c("ASV", "comparison")) %>% 
  #left_join(res_diff, by = c("ASV", "comparison")) %>% 
  left_join(df_taxa, by = "ASV") %>% 
  mutate(comparison = "Il_5wo-vs-Il_7wo") %>% 
  mutate(reference_group = "Il_7wo") %>% 
  mutate(comparison_group = "Il_5wo") %>% 
  bind_rows(res_long)

```




```{r}
# Ce_5wo vs Ce_7wo

physeq_subset <- physeq %>% 
  subset_samples(diet == "NoAbx" & site == "CS") 

physeq_subset <- prune_taxa(taxa_sums(physeq_subset) > 0, physeq_subset) 

#physeq_subset <- phyloseq_filter_prevalence(physeq_subset, prev.trh = 0.0, abund.trh = 5, threshold_condition = "AND", abund.type = "mean")

physeq_subset

phyloseq_to_df(physeq_subset, addtax = F) %>% 
  pivot_longer(!OTU, names_to = "sample_names", values_to = "counts") %>% 
  left_join(df_sample, by = "sample_names") %>% 
  select(sample_names, site_necropsy) %>% 
  distinct(site_necropsy)



out <- ancombc(phyloseq = physeq_subset, formula = "site_necropsy", 
              p_adj_method = "fdr", zero_cut = 0.90, lib_cut = 0, 
              struc_zero = FALSE, neg_lb = TRUE, tol = 1e-5, 
              max_iter = 100, conserve = FALSE, alpha = 0.05, global = FALSE)

res <- out$res

res_beta <- res$beta %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "beta")

res_se <- res$se %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "se")

res_w <- res$W %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "W")

res_pval <- res$p_val %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "p_val")

res_qval <- res$q_val %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "q_val")

res_diff <- res$diff_abn %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "diff_abn")

res_long <- res_beta %>% 
  left_join(res_se, by = c("ASV", "comparison")) %>% 
  left_join(res_w, by = c("ASV", "comparison")) %>% 
  left_join(res_pval, by = c("ASV", "comparison")) %>% 
  #left_join(res_qval, by = c("ASV", "comparison")) %>% 
  #left_join(res_diff, by = c("ASV", "comparison")) %>% 
  left_join(df_taxa, by = "ASV") %>% 
  mutate(comparison = "Ce_5wo-vs-Ce_7wo") %>% 
  mutate(reference_group = "Ce_7wo") %>% 
  mutate(comparison_group = "Ce_5wo") %>% 
  bind_rows(res_long)





```




```{r}
# -Jej_5wo vs Il_5wo


physeq_subset <- physeq %>% 
  subset_samples(diet == "NoAbx" & site != "CS" & site != "CC" & necropsy == "Two_wpw") 

physeq_subset <- prune_taxa(taxa_sums(physeq_subset) > 0, physeq_subset) 

#physeq_subset <- phyloseq_filter_prevalence(physeq_subset, prev.trh = 0.0, abund.trh = 5, threshold_condition = "AND", abund.type = "mean")

physeq_subset

phyloseq_to_df(physeq_subset, addtax = F) %>% 
  pivot_longer(!OTU, names_to = "sample_names", values_to = "counts") %>% 
  left_join(df_sample, by = "sample_names") %>% 
  select(sample_names, site_necropsy) %>% 
  distinct(site_necropsy)


out <- ancombc(phyloseq = physeq_subset, formula = "site_necropsy", 
              p_adj_method = "fdr", zero_cut = 0.90, lib_cut = 0, 
              struc_zero = FALSE, neg_lb = TRUE, tol = 1e-5, 
              max_iter = 100, conserve = FALSE, alpha = 0.05, global = FALSE)

res <- out$res

res_beta <- res$beta %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "beta")

res_se <- res$se %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "se")

res_w <- res$W %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "W")

res_pval <- res$p_val %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "p_val")

res_qval <- res$q_val %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "q_val")

res_diff <- res$diff_abn %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "diff_abn")

res_long <- res_beta %>% 
  left_join(res_se, by = c("ASV", "comparison")) %>% 
  left_join(res_w, by = c("ASV", "comparison")) %>% 
  left_join(res_pval, by = c("ASV", "comparison")) %>% 
  #left_join(res_qval, by = c("ASV", "comparison")) %>% 
  #left_join(res_diff, by = c("ASV", "comparison")) %>% 
  left_join(df_taxa, by = "ASV") %>% 
  mutate(comparison = "Jej_5wo-vs-Il_5wo") %>% 
  mutate(reference_group = "Il_5wo") %>% 
  mutate(comparison_group = "Jej_5wo") %>% 
  bind_rows(res_long)

```



```{r}
# -Jej_5wo vs Ce_5wo


physeq_subset <- physeq %>% 
  subset_samples(diet == "NoAbx" & site != "Il" & site != "CC" & necropsy == "Two_wpw") 

physeq_subset <- prune_taxa(taxa_sums(physeq_subset) > 0, physeq_subset) 

#physeq_subset <- phyloseq_filter_prevalence(physeq_subset, prev.trh = 0.0, abund.trh = 5, threshold_condition = "AND", abund.type = "mean")

physeq_subset

phyloseq_to_df(physeq_subset, addtax = F) %>% 
  pivot_longer(!OTU, names_to = "sample_names", values_to = "counts") %>% 
  left_join(df_sample, by = "sample_names") %>% 
  select(sample_names, site_necropsy) %>% 
  distinct(site_necropsy)


out <- ancombc(phyloseq = physeq_subset, formula = "site_necropsy", 
              p_adj_method = "fdr", zero_cut = 0.90, lib_cut = 0, 
              struc_zero = FALSE, neg_lb = TRUE, tol = 1e-5, 
              max_iter = 100, conserve = FALSE, alpha = 0.05, global = FALSE)

res <- out$res

res_beta <- res$beta %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "beta")

res_se <- res$se %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "se")

res_w <- res$W %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "W")

res_pval <- res$p_val %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "p_val")

res_qval <- res$q_val %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "q_val")

res_diff <- res$diff_abn %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "diff_abn")

res_long <- res_beta %>% 
  left_join(res_se, by = c("ASV", "comparison")) %>% 
  left_join(res_w, by = c("ASV", "comparison")) %>% 
  left_join(res_pval, by = c("ASV", "comparison")) %>% 
  #left_join(res_qval, by = c("ASV", "comparison")) %>% 
  #left_join(res_diff, by = c("ASV", "comparison")) %>% 
  left_join(df_taxa, by = "ASV") %>% 
  mutate(comparison = "Jej_5wo-vs-Ce_5wo") %>% 
  mutate(reference_group = "Ce_5wo") %>% 
  mutate(comparison_group = "Jej_5wo") %>% 
  bind_rows(res_long)

```



```{r}
# -Il_5wo vs Ce_5w0


physeq_subset <- physeq %>% 
  subset_samples(diet == "NoAbx" & site != "Jej" & site != "CC" & necropsy == "Two_wpw") 

physeq_subset <- prune_taxa(taxa_sums(physeq_subset) > 0, physeq_subset) 

#physeq_subset <- phyloseq_filter_prevalence(physeq_subset, prev.trh = 0.0, abund.trh = 5, threshold_condition = "AND", abund.type = "mean")

physeq_subset

phyloseq_to_df(physeq_subset, addtax = F) %>% 
  pivot_longer(!OTU, names_to = "sample_names", values_to = "counts") %>% 
  left_join(df_sample, by = "sample_names") %>% 
  select(sample_names, site_necropsy) %>% 
  distinct(site_necropsy)


out <- ancombc(phyloseq = physeq_subset, formula = "site_necropsy", 
              p_adj_method = "fdr", zero_cut = 0.90, lib_cut = 0, 
              struc_zero = FALSE, neg_lb = TRUE, tol = 1e-5, 
              max_iter = 100, conserve = FALSE, alpha = 0.05, global = FALSE)

res <- out$res

res_beta <- res$beta %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "beta")

res_se <- res$se %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "se")

res_w <- res$W %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "W")

res_pval <- res$p_val %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "p_val")

res_qval <- res$q_val %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "q_val")

res_diff <- res$diff_abn %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "diff_abn")

res_long <- res_beta %>% 
  left_join(res_se, by = c("ASV", "comparison")) %>% 
  left_join(res_w, by = c("ASV", "comparison")) %>% 
  left_join(res_pval, by = c("ASV", "comparison")) %>% 
  #left_join(res_qval, by = c("ASV", "comparison")) %>% 
  #left_join(res_diff, by = c("ASV", "comparison")) %>% 
  left_join(df_taxa, by = "ASV") %>% 
  mutate(comparison = "Il_5wo-vs-Ce_5w0") %>% 
  mutate(reference_group = "Ce_5wo") %>% 
  mutate(comparison_group = "Il_5wo") %>% 
  bind_rows(res_long)

```



```{r}
# # -Jej_7wo vs Il_7wo


physeq_subset <- physeq %>% 
  subset_samples(diet == "NoAbx" & site != "CS" & site != "CC" & necropsy == "Four_wpw") 

physeq_subset <- prune_taxa(taxa_sums(physeq_subset) > 0, physeq_subset) 

#physeq_subset <- phyloseq_filter_prevalence(physeq_subset, prev.trh = 0.0, abund.trh = 5, threshold_condition = "AND", abund.type = "mean")

physeq_subset

phyloseq_to_df(physeq_subset, addtax = F) %>% 
  pivot_longer(!OTU, names_to = "sample_names", values_to = "counts") %>% 
  left_join(df_sample, by = "sample_names") %>% 
  select(sample_names, site_necropsy) %>% 
  distinct(site_necropsy)


out <- ancombc(phyloseq = physeq_subset, formula = "site_necropsy", 
              p_adj_method = "fdr", zero_cut = 0.90, lib_cut = 0, 
              struc_zero = FALSE, neg_lb = TRUE, tol = 1e-5, 
              max_iter = 100, conserve = FALSE, alpha = 0.05, global = FALSE)

res <- out$res

res_beta <- res$beta %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "beta")

res_se <- res$se %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "se")

res_w <- res$W %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "W")

res_pval <- res$p_val %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "p_val")

res_qval <- res$q_val %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "q_val")

res_diff <- res$diff_abn %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "diff_abn")

res_long <- res_beta %>% 
  left_join(res_se, by = c("ASV", "comparison")) %>% 
  left_join(res_w, by = c("ASV", "comparison")) %>% 
  left_join(res_pval, by = c("ASV", "comparison")) %>% 
  #left_join(res_qval, by = c("ASV", "comparison")) %>% 
  #left_join(res_diff, by = c("ASV", "comparison")) %>% 
  left_join(df_taxa, by = "ASV") %>% 
  mutate(comparison = "Jej_7wo-vs-Il_7wo") %>% 
  mutate(reference_group = "Il_7wo") %>% 
  mutate(comparison_group = "Jej_7wo") %>% 
  bind_rows(res_long)

```


```{r}
# -Jej_7wo vs Ce_7wo


physeq_subset <- physeq %>% 
  subset_samples(diet == "NoAbx" & site != "Il" & site != "CC" & necropsy == "Four_wpw") 

physeq_subset <- prune_taxa(taxa_sums(physeq_subset) > 0, physeq_subset) 

#physeq_subset <- phyloseq_filter_prevalence(physeq_subset, prev.trh = 0.0, abund.trh = 5, threshold_condition = "AND", abund.type = "mean")

physeq_subset

phyloseq_to_df(physeq_subset, addtax = F) %>% 
  pivot_longer(!OTU, names_to = "sample_names", values_to = "counts") %>% 
  left_join(df_sample, by = "sample_names") %>% 
  select(sample_names, site_necropsy) %>% 
  distinct(site_necropsy)


out <- ancombc(phyloseq = physeq_subset, formula = "site_necropsy", 
              p_adj_method = "fdr", zero_cut = 0.90, lib_cut = 0, 
              struc_zero = FALSE, neg_lb = TRUE, tol = 1e-5, 
              max_iter = 100, conserve = FALSE, alpha = 0.05, global = FALSE)

res <- out$res

res_beta <- res$beta %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "beta")

res_se <- res$se %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "se")

res_w <- res$W %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "W")

res_pval <- res$p_val %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "p_val")

res_qval <- res$q_val %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "q_val")

res_diff <- res$diff_abn %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "diff_abn")

res_long <- res_beta %>% 
  left_join(res_se, by = c("ASV", "comparison")) %>% 
  left_join(res_w, by = c("ASV", "comparison")) %>% 
  left_join(res_pval, by = c("ASV", "comparison")) %>% 
  #left_join(res_qval, by = c("ASV", "comparison")) %>% 
  #left_join(res_diff, by = c("ASV", "comparison")) %>% 
  left_join(df_taxa, by = "ASV") %>% 
  mutate(comparison = "Jej_7wo-vs-Ce_7wo") %>% 
  mutate(reference_group = "Ce_7wo") %>% 
  mutate(comparison_group = "Jej_7wo") %>% 
  bind_rows(res_long)

```



```{r}
# -Il_7wo vs Ce_7w0


physeq_subset <- physeq %>% 
  subset_samples(diet == "NoAbx" & site != "Jej" & site != "CC" & necropsy == "Four_wpw") 

physeq_subset <- prune_taxa(taxa_sums(physeq_subset) > 0, physeq_subset) 

#physeq_subset <- phyloseq_filter_prevalence(physeq_subset, prev.trh = 0.0, abund.trh = 5, threshold_condition = "AND", abund.type = "mean")

physeq_subset

phyloseq_to_df(physeq_subset, addtax = F) %>% 
  pivot_longer(!OTU, names_to = "sample_names", values_to = "counts") %>% 
  left_join(df_sample, by = "sample_names") %>% 
  select(sample_names, site_necropsy) %>% 
  distinct(site_necropsy)


out <- ancombc(phyloseq = physeq_subset, formula = "site_necropsy", 
              p_adj_method = "fdr", zero_cut = 0.90, lib_cut = 0, 
              struc_zero = FALSE, neg_lb = TRUE, tol = 1e-5, 
              max_iter = 100, conserve = FALSE, alpha = 0.05, global = FALSE)

res <- out$res

res_beta <- res$beta %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "beta")

res_se <- res$se %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "se")

res_w <- res$W %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "W")

res_pval <- res$p_val %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "p_val")

res_qval <- res$q_val %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "q_val")

res_diff <- res$diff_abn %>% 
  rownames_to_column(var = "ASV") %>% 
  pivot_longer(!ASV, names_to = "comparison", values_to = "diff_abn")

res_long <- res_beta %>% 
  left_join(res_se, by = c("ASV", "comparison")) %>% 
  left_join(res_w, by = c("ASV", "comparison")) %>% 
  left_join(res_pval, by = c("ASV", "comparison")) %>% 
  #left_join(res_qval, by = c("ASV", "comparison")) %>% 
  #left_join(res_diff, by = c("ASV", "comparison")) %>% 
  left_join(df_taxa, by = "ASV") %>% 
  mutate(comparison = "Il_7wo-vs-Ce_7wo") %>%
  mutate(reference_group = "Ce_7wo") %>% 
  mutate(comparison_group = "Il_7wo") %>% 
  bind_rows(res_long)

```



```{r}
# FDR correction for multiple testing
# holm is default and suggested to keep FDR at 5% value
# Benjamini & Hochberg (1995) ("BH" or its alias "fdr") - gives more power but increases FDR


res_long$q_val <- p.adjust(res_long$p_val, method = "holm", n = length(res_long$p_val))

print(dim(res_long))

res_long_corrected <- res_long %>% 
  select(ASV, comparison, reference_group, comparison_group, beta, se, W, p_val, q_val, domain, phylum, class, order, family, genus) %>% 
  mutate(p_val = round(p_val, 10)) %>% 
  mutate(q_val = round(q_val, 10))


res_long_corrected %>% 
  filter(q_val < 0.05) %>% 
  count(comparison)


# As shown in the user manual, W is the standardized log fold-change (beta), which is a Wald-type of statistic and the p-value is computed by assuming Nomarlity.

# negative beta/W, means higher abundance in reference group
# positive beta/W, means higher abundance in comparison group

write_tsv(res_long_corrected, "diff_abundance_all_comparisons_no_ab_filter.tsv")

```



